File: apps/core/src/core/schemas/content.py
"""Trash schemas: `GET /content/trash` + `POST /content/restore`.

Also the unified feed: `GET /content`.
"""

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field

ResourceType = Literal["layer", "project", "folder", "bundle", "template"]
ContentView = Literal["space", "shared_with_me", "shared_with_space", "recent"]
# The sort keys and directions the feed accepts. `crud_content._validate_order`
# holds the same allow-list and maps each to the SQL it formats into the
# query; these types reject anything else at the request boundary (422)
# before it ever gets there.
ContentOrderBy = Literal["updated_at", "created_at", "name", "last_opened_at"]
ContentOrder = Literal["ascendent", "descendent"]


class TrashItem(BaseModel):
    """One restorable entry in the trash listing.

    A trashed item whose own folder is also trashed is not listed
    separately — it is folded into that folder's entry (restoring the
    folder restores its whole subtree). See `CRUDTrash.list`.
    """

    type: ResourceType
    id: UUID
    name: str
    deleted_at: datetime
    purge_after: datetime = Field(
        description="When the purge task removes this item for good "
        "(30 days after `deleted_at`)."
    )


class RestoreItem(BaseModel):
    type: ResourceType
    id: UUID


class RestoreRequest(BaseModel):
    items: list[RestoreItem]


class ShareEntry(BaseModel):
    """One direct grant on a resource, as `fetch_grants_by_resource`
    (core/content.py) buckets it — a team/organisation/user grantee shaped
    identically so a caller can render all three kinds the same way."""

    role: str
    id: UUID
    name: str | None = None
    avatar: str | None = None


class SharedWith(BaseModel):
    """Every direct grant on one resource, bucketed by grantee kind. Mirrors
    the dict shape `fetch_grants_by_resource` returns per resource id."""

    teams: list[ShareEntry] = Field(default_factory=list)
    organizations: list[ShareEntry] = Field(default_factory=list)
    users: list[ShareEntry] = Field(default_factory=list)


class ContentCreator(BaseModel):
    """The person who created a feed row — `user_id` on the content table,
    resolved to what a card or row can draw: a display name and a picture."""

    id: UUID
    name: str
    avatar: str | None = None


class ContentItem(BaseModel):
    """One row in the unified feed (`GET /content`) — a folder, project,
    layer or bundle the caller may read."""

    type: ResourceType
    id: UUID
    name: str
    space_id: UUID
    folder_id: UUID | None = Field(
        default=None,
        description="Parent folder id — a folder's own parent for a "
        "`type: folder` row, the containing folder for anything else.",
    )
    updated_at: datetime
    created_at: datetime
    my_role: Literal["owner", "editor", "viewer"]
    created_by: ContentCreator | None = Field(
        default=None,
        description="Who created the row. None when the creating account no "
        "longer exists.",
    )
    shared_with: SharedWith | None = None
    thumbnail_url: str | None = Field(
        default=None,
        description="A URL the client can load: a presigned link for a "
        "stored thumbnail, or the default artwork for this row's kind.",
    )
    is_public: bool = Field(
        default=False,
        description="A published public snapshot exists for this row "
        "(projects only; always False for the other kinds).",
    )
    layer_type: str | None = None
    feature_layer_geometry_type: str | None = None
    is_shortcut: bool = Field(
        default=False,
        description="True for a `content_shortcut` row left behind after a "
        "transfer — a badged pointer at an item that now lives elsewhere, "
        "shown with that item's real name/type/thumbnail.",
    )
    restricted: bool = Field(
        default=False,
        description="Restricted (D9): members of this row's space do not get "
        "the space default role on it — only its space owner/admin and "
        "whoever holds a grant reach it.",
    )
    restricted_inherited: bool = Field(
        default=False,
        description="Restricted through an ancestor folder, or through a "
        "bundle holding this layer; toggled there, not here.",
    )
    template_payload_kind: Literal["workflow", "layout", "project"] | None = Field(
        default=None,
        description="`type: template` rows only: what the template's config "
        "carries (T1). None for every other row.",
    )
    template_kinds: list[Literal["workflow", "dashboard", "layout"]] = Field(
        default_factory=list,
        description="`type: template` rows only: the kinds shown on the "
        "card, derived from `template_payload_kind` (T1). Empty for every "
        "other row.",
    )
    template_catalog_status: (
        Literal["none", "proposed", "published", "declined"] | None
    ) = Field(
        default=None,
        description="`type: template` rows only: the GOAT catalog shelf "
        "state (T4). None for every other row.",
    )
    template_ships_sample_data: bool = Field(
        default=False,
        description="`type: template` rows only: true once any declared "
        "input (T5) is a shipped, catalog-origin dataset. Always False for "
        "every other row.",
    )


class RestrictedUpdate(BaseModel):
    restricted: bool = Field(
        description="True closes the folder or item off from the space "
        "default role; False lifts it."
    )


class ContentPage(BaseModel):
    items: list[ContentItem]
    total: int
    page: int
    size: int

File: apps/web/lib/validations/content.ts
import * as z from "zod";

export const spaceKind = z.enum(["personal", "team", "organization"]);
export const spaceDefaultRole = z.enum(["viewer", "editor"]);
export const contentRole = z.enum(["owner", "editor", "viewer"]);
export const contentType = z.enum(["layer", "project", "folder", "bundle", "template"]);
export const contentView = z.enum(["space", "shared_with_me", "shared_with_space", "recent"]);

export const spaceSchema = z.object({
  id: z.string().uuid(),
  kind: spaceKind,
  name: z.string(),
  default_role: spaceDefaultRole,
  my_role: contentRole.nullable().optional(),
  team_id: z.string().uuid().nullable().optional(),
  organization_id: z.string().uuid().nullable().optional(),
});

export const shareEntrySchema = z.object({
  role: z.string(),
  id: z.string().uuid(),
  name: z.string().nullable().optional(),
  avatar: z.string().nullable().optional(),
});

export const sharedWithSchema = z.object({
  teams: z.array(shareEntrySchema).default([]),
  organizations: z.array(shareEntrySchema).default([]),
  users: z.array(shareEntrySchema).default([]),
});

/** Who created a feed row — the content table's `user_id`, resolved to a
 * display name and picture. Null once the creating account is gone. */
export const contentCreatorSchema = z.object({
  id: z.string().uuid(),
  name: z.string(),
  avatar: z.string().nullable().optional(),
});

export const contentItemSchema = z.object({
  type: contentType,
  id: z.string().uuid(),
  name: z.string(),
  space_id: z.string().uuid(),
  folder_id: z.string().uuid().nullable().optional(),
  updated_at: z.string(),
  created_at: z.string(),
  my_role: contentRole,
  created_by: contentCreatorSchema.nullable().optional(),
  shared_with: sharedWithSchema.nullable().default(null),
  thumbnail_url: z.string().nullable().optional(),
  /** A published public snapshot exists — projects only. */
  is_public: z.boolean().default(false),
  layer_type: z.string().nullable().optional(),
  feature_layer_geometry_type: z.string().nullable().optional(),
  is_shortcut: z.boolean().default(false),
  restricted: z.boolean().default(false),
  restricted_inherited: z.boolean().default(false),
  /** `type: "template"` rows only: what the template's config carries (T1). */
  template_payload_kind: z.enum(["workflow", "layout", "project"]).nullable().optional(),
  /** `type: "template"` rows only: the kinds shown on the card (T1). Left
   * optional (rather than defaulted) so the countless existing `ContentItem`
   * fixtures across the codebase — folders, projects, layers, bundles, none
   * of which ever carry it — do not all need updating for a field that is
   * only ever populated on a template row. Read as `item.template_kinds ?? []`. */
  template_kinds: z.array(z.enum(["workflow", "dashboard", "layout"])).optional(),
  /** `type: "template"` rows only: the GOAT catalog shelf state (T4). */
  template_catalog_status: z.enum(["none", "proposed", "published", "declined"]).nullable().optional(),
  /** `type: "template"` rows only: true once any declared input is a
   * shipped, catalog-origin dataset (T5). Optional for the same reason as
   * `template_kinds` above — read as `item.template_ships_sample_data ?? false`. */
  template_ships_sample_data: z.boolean().optional(),
});

export const contentPageSchema = z.object({
  items: z.array(contentItemSchema),
  total: z.number().int(),
  page: z.number().int(),
  size: z.number().int(),
});

export const contentQueryParamsSchema = z.object({
  view: contentView,
  space_id: z.string().uuid().optional(),
  folder_id: z.string().uuid().optional(),
  types: z.string().optional(),
  search: z.string().optional(),
  order_by: z.enum(["updated_at", "created_at", "name", "last_opened_at"]).optional(),
  order: z.enum(["ascendent", "descendent"]).optional(),
  page: z.number().int().positive().optional(),
  size: z.number().int().positive().optional(),
});

export const trashItemSchema = z.object({
  type: contentType,
  id: z.string().uuid(),
  name: z.string(),
  deleted_at: z.string(),
  purge_after: z.string(),
});

export const contentRefSchema = z.object({ type: contentType, id: z.string() });

export const transferPreviewRequestSchema = z.object({
  items: z.array(contentRefSchema),
  target_space_id: z.string(),
});

export const transferRequestSchema = transferPreviewRequestSchema.extend({
  dataset_ids: z.array(z.string()),
  leave_shortcut: z.boolean(),
});

export const transferPreviewSchema = z.object({
  items: z.array(contentRefSchema.extend({ name: z.string() })),
  datasets: z.array(
    z.object({ id: z.string(), name: z.string(), owned: z.boolean(), used_elsewhere: z.boolean() })
  ),
  grants_to_drop: z.number().int(),
  folders_in_subtrees: z.number().int(),
  trashed_in_subtrees: z.number().int(),
  skipped_foreign: z.number().int().default(0),
  name_collisions: z.array(z.string()).default([]),
  warnings: z.array(z.string()),
});

export const transferResultSchema = z.object({
  moved: z.object({
    folder: z.number().int().default(0),
    project: z.number().int().default(0),
    layer: z.number().int().default(0),
    bundle: z.number().int().default(0),
  }),
  shortcuts: z.number().int(),
  trashed_moved: z.number().int(),
});

/** `GET /space/{space_id}/usage` (D13, members only): total size and count
 * of the space's live layers, and count of its live projects. */
export const spaceUsageSchema = z.object({
  space_id: z.string().uuid(),
  bytes: z.number().int(),
  layers: z.number().int(),
  projects: z.number().int(),
});

export type SpaceKind = z.infer<typeof spaceKind>;
export type SpaceDefaultRole = z.infer<typeof spaceDefaultRole>;
export type ContentRole = z.infer<typeof contentRole>;
export type ContentType = z.infer<typeof contentType>;
export type ContentView = z.infer<typeof contentView>;
export type Space = z.infer<typeof spaceSchema>;
export type ShareEntry = z.infer<typeof shareEntrySchema>;
export type SharedWith = z.infer<typeof sharedWithSchema>;
export type ContentCreator = z.infer<typeof contentCreatorSchema>;
export type ContentItem = z.infer<typeof contentItemSchema>;
export type ContentPage = z.infer<typeof contentPageSchema>;
export type ContentQueryParams = z.infer<typeof contentQueryParamsSchema>;
export type TrashItem = z.infer<typeof trashItemSchema>;
export type ContentRef = z.infer<typeof contentRefSchema>;
export type TransferPreviewRequest = z.infer<typeof transferPreviewRequestSchema>;
export type TransferRequest = z.infer<typeof transferRequestSchema>;
export type TransferPreview = z.infer<typeof transferPreviewSchema>;
export type TransferResult = z.infer<typeof transferResultSchema>;
export type SpaceUsage = z.infer<typeof spaceUsageSchema>;
